geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/SimpleFeatureSerializer.scala

```diff
@@ -0,0 +1,346 @@
+    }
+
+  /**
+   * Finds an decoding function based on the input type
+   *
+   * @param clas
+   * @param metadata
+   * @return
+   */
+  def matchDecode(clas: Class[_], metadata: jMap[Object, Object]): (Input) => AnyRef = clas match {
+
+    case c if classOf[String].isAssignableFrom(c) =>
+      (in: Input) => in.readString()
+
+    case c if classOf[java.lang.Integer].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readInt().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Long].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readLong().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Double].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readDouble().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Float].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readFloat().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Boolean].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readBoolean().asInstanceOf[AnyRef]
+
+    case c if classOf[Date].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else new Date(in.readLong())
+
+    case c if classOf[UUID].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) {
+        null
+      } else {
+        val mostSignificantBits = in.readLong()
+        val leastSignificantBits = in.readLong()
+        new UUID(mostSignificantBits, leastSignificantBits)
+      }
+
+    case c if classOf[Geometry].isAssignableFrom(c) =>
+      (in: Input) => {
+        val length = in.readInt(true)
+        if (length > 0) {
+          val bytes = new Array[Byte](length)
+          in.read(bytes)
+          WKBUtils.read(bytes)
+        } else {
+          null
+        }
+      }
+
+    case c if classOf[jList[_]].isAssignableFrom(c) =>
+      val subtype = metadata.get("subtype").asInstanceOf[Class[_]]
+      val subDecoding = matchDecode(subtype, null)
+
+      (in: Input) => {
+        val length = in.readInt()
+        if (length < 0) {
+          null
+        } else {
+          val list = new java.util.ArrayList[Object](length)
+          (0 until length).foreach(_ => list.add(subDecoding(in)))
+          list
+        }
+      }
+
+    case c if classOf[jMap[_, _]].isAssignableFrom(c) =>
+      val keyClass      = metadata.get("keyclass").asInstanceOf[Class[_]]
+      val valueClass    = metadata.get("valueclass").asInstanceOf[Class[_]]
+      val keyDecoding   = matchDecode(keyClass, null)
+      val valueDecoding = matchDecode(valueClass, null)
+
+      (in: Input) => {
+        val length = in.readInt()
+        if (length < 0) {
+          null
+        } else {
+          val map = new java.util.HashMap[Object, Object](length)
+          (0 until length).foreach(_ => map.put(keyDecoding(in), valueDecoding(in)))
```

Comment:
From a cursory look on the interwebs, foreach in Scala is usually much slower than the similar while loop.  

Since we are in a race for speed, I'd consider ditching 'inner' foreachs...